File: apps/flutter_app/lib/models/guest.dart
enum JoinRequestStatus { pending, approved, rejected }

class Guest {
  final String id;
  final String eventId;
  final String displayName;
  final String? phone;
  final DateTime joinedAt;
  final DateTime? kickedAt;
  final int photoCount;

  Guest({
    required this.id,
    required this.eventId,
    required this.displayName,
    this.phone,
    required this.joinedAt,
    this.kickedAt,
    this.photoCount = 0,
  });

  factory Guest.fromJson(Map<String, dynamic> json) => Guest(
    id: json['id'],
    eventId: json['event_id'],
    displayName: json['display_name'],
    phone: json['phone'],
    joinedAt: DateTime.parse(json['joined_at']),
    kickedAt: json['kicked_at'] != null ? DateTime.parse(json['kicked_at']) : null,
    photoCount: json['photo_count'] ?? 0,
  );
}

class JoinRequest {
  final String id;
  final String eventId;
  final String displayName;
  final String? phone;
  final JoinRequestStatus status;
  final DateTime requestedAt;

  JoinRequest({
    required this.id,
    required this.eventId,
    required this.displayName,
    this.phone,
    required this.status,
    required this.requestedAt,
  });

  factory JoinRequest.fromJson(Map<String, dynamic> json) => JoinRequest(
    id: json['id'],
    eventId: json['event_id'],
    displayName: json['display_name'],
    phone: json['phone'],
    status: JoinRequestStatus.values.byName(json['status']),
    requestedAt: DateTime.parse(json['requested_at']),
  );
}

File: apps/api/app/models/guest.py
import uuid
from datetime import datetime

from sqlalchemy import Boolean, String, DateTime, ForeignKey, Enum as SAEnum
from sqlalchemy.orm import Mapped, mapped_column, relationship

from app.database import Base
from app.schemas import JoinRequestStatus


class Guest(Base):
    __tablename__ = "event_guests"

    id: Mapped[str] = mapped_column(String, primary_key=True, default=lambda: uuid.uuid4().hex[:12])
    event_id: Mapped[str] = mapped_column(String, ForeignKey("events.id"))
    display_name: Mapped[str] = mapped_column(String(100))
    phone: Mapped[str | None] = mapped_column(String(20), nullable=True)
    # Marks the host's own "camera row". The host can also contribute photos
    # to their own event — we model that as a Guest tied to their user_id so
    # the rest of the upload pipeline works unchanged. is_host is the UI hint
    # so the album can badge them differently.
    is_host: Mapped[bool] = mapped_column(Boolean, default=False)
    user_id: Mapped[str | None] = mapped_column(String, nullable=True, index=True)
    joined_at: Mapped[datetime] = mapped_column(DateTime(timezone=True), default=datetime.utcnow)
    kicked_at: Mapped[datetime | None] = mapped_column(DateTime(timezone=True), nullable=True)

    event = relationship("Event", back_populates="guests")
    photos = relationship("Photo", back_populates="guest", lazy="dynamic")


class JoinRequest(Base):
    __tablename__ = "join_requests"

    id: Mapped[str] = mapped_column(String, primary_key=True, default=lambda: uuid.uuid4().hex[:12])
    event_id: Mapped[str] = mapped_column(String, ForeignKey("events.id"))
    display_name: Mapped[str] = mapped_column(String(100))
    phone: Mapped[str | None] = mapped_column(String(20), nullable=True)
    status: Mapped[JoinRequestStatus] = mapped_column(SAEnum(JoinRequestStatus), default=JoinRequestStatus.pending)
    requested_at: Mapped[datetime] = mapped_column(DateTime(timezone=True), default=datetime.utcnow)
